src/globus_cli/parsing/param_types/json_strorfile.py

```diff
@@ -0,0 +1,107 @@
+from __future__ import annotations
+
+import dataclasses
+import json
+import os
+import sys
+import typing as t
+
+import click
+
+from globus_cli.constants import EXPLICIT_NULL, ExplicitNullType
+from globus_cli.types import JsonValue
+
+from .annotated_param import AnnotatedParamType
+
+
+@dataclasses.dataclass
+class ParsedJSONData:
+    # str means it's the path to a file input
+    # including `-` for stdin
+    # None means the parse came from an argument input
+    filename: t.Union[str, None]
+    # data is the parsed data
+    data: JsonValue
+
+
+class JSONStringOrFile(AnnotatedParamType):
+    """
+    Parse an input which could be a filename or could be a JSON blob being
+    supplied on the commandline.
+
+    For example,
+        --opt './foo/bar.json'
+        --opt "$(./myscript.sh)"
+
+    For compatibility with the older mode of use, this param also supports a
+    `file:` prefix which disambiguates an input as being a filename. e.g.
+        --opt 'file:./foo/bar.json'
+
+    The parsing happens in the following order of precedence:
+        - 'file:' prefix (filename)
+        - bare filename, including `-` for stdin
+        - JSON parse
+
+    This ordering protects against odd behavior if a filename happens to be valid JSON.
+    Most likely to happen with a JSON literal like an integer, e.g. `--opt 100`
```

Comment:
Based on my own past experience, I would recommend changing this example to `--opt 1`, which seems like a more likely filename if STDERR is piped incorrectly to STDOUT and results in piping to a file literally named `1`.

```suggestion
    Most likely to happen with a JSON literal like an integer, e.g. `--opt 1`
```